BGAnimations/ScreenSelectMusic overlay/PerPlayer/DensityGraph.lua

```diff
@@ -0,0 +1,132 @@
+local player = ...
+local pn = ToEnumShortString(player)
+
+-- Height and width of the density graph.
+local height = 64
+local width = IsUsingWideScreen() and 286 or 276
+
+local af = Def.ActorFrame{
+	InitCommand=function(self)
+		self:visible( GAMESTATE:IsHumanPlayer(player) )
+		self:xy(_screen.cx-182, _screen.cy+23)
+
+		if player == PLAYER_2 then
+			self:addy(height+24)
+		end
+
+		if IsUsingWideScreen() then
+			self:addx(-5)
+		end
+	end,
+	PlayerJoinedMessageCommand=function(self, params)
+		if params.Player == player then
+			self:visible(true)
+		end
+	end,
+	PlayerUnjoinedMessageCommand=function(self, params)
+		if params.Player == player then
+			self:visible(false)
+		end
+	end,
+	CurrentSongChangedMessageCommand=function(self) self:queuecommand("UpdateGraph") end,
+	["CurrentSteps"..pn.."ChangedMessageCommand"]=function(self) self:queuecommand("UpdateGraph") end,
+
+	UpdateGraphCommand=function(self)
+		if not GAMESTATE:IsCourseMode() and GAMESTATE:GetCurrentSong() then
+			self:GetChild("Breakdown"):visible(true)
+			self:GetChild("DensityGraph"):visible(true)
+			self:GetChild("NPS"):visible(false)
+		else
+			self:GetChild("Breakdown"):visible(false)
+			self:GetChild("DensityGraph"):visible(false)
+			self:GetChild("NPS"):settext("Peak NPS: ")
+			self:GetChild("NPS"):visible(false)
+		end
+	end,
+}
+
+-- Background quad for the density graph
+af[#af+1] = Def.Quad{
+	InitCommand=function(self)
+		self:diffuse(color("#1e282f")):zoomto(width, height)
+		if ThemePrefs.Get("RainbowMode") then
+			self:diffusealpha(0.9)
+		end
+	end
+}
+
+-- The Density Graph itself
+-- The name "DensityGraph_AMV" comes from SL-Histogram.lua
+af[#af+1] = NPS_Histogram(player, width, height)..{
+	Name="DensityGraph",
+	OnCommand=function(self)
+		self:addx(-width/2):addy(height/2)
+	end,
+}
+
+-- The Peak NPS text
+af[#af+1] = LoadFont("Miso/_miso")..{
+	Name="NPS",
+	Text="Peak NPS: ",
+	InitCommand=function(self)
+		self:horizalign(left):zoom(0.8)
+		if player == PLAYER_1 then
+			self:addx(60):addy(-43)
+		else
+			self:addx(-136):addy(-43)
+		end
+	end,
+	PeakNPSUpdatedMessageCommand=function(self)
+		local peakNps = nil
+		if player == PLAYER_1 then
+			peakNps = GAMESTATE:Env()["P1PeakNPS"]
+		else
+			peakNps = GAMESTATE:Env()["P2PeakNPS"]
+		end
+		if peakNps then
+			self:settext(("Peak NPS: %.1f"):format(peakNps))
+		end
+	end
+}
+
+-- Breakdown
+af[#af+1] = Def.ActorFrame{
+	Name="Breakdown",
+	InitCommand=function(self)
+		local actor_height = 17
+		self:addy(height/2 - actor_height/2)
+	end,
+
+	Def.Quad{
+		InitCommand=function(self)
+			local bg_height = 17
+			self:diffuse(color("#000000")):zoomto(width, bg_height):diffusealpha(0.5)
+		end
+	},
+	
+	LoadFont("Miso/_miso")..{
+		Text="No Streams!",
```

Comment:
For the future: I think it would be nice if no text and background quad would be displayed if we actually didn't find any streams in the chart.

Reply:
Agreed! My plan right now is to display streams if it's a stream chart, display tech stuff if it's a tech chart, or hide this altogether if it's not deemed as either.